rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp

```diff
@@ -0,0 +1,335 @@
+        continue;
+      }
+      group->find_service_ptrs_if([&exec_list](const rclcpp::ServiceBase::SharedPtr & service) {
+          if(service){
+            // If any service is found, push it in the exec_list struct
+            exec_list.service.push_back(service);
+            exec_list.number_of_services++;
+          }
+          return false;
+        });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_client_list(ExecutableList & exec_list)
+{
+  // Clear the previous clients (if any) from the ExecutableList struct
+  exec_list.client.clear();
+  exec_list.number_of_clients = 0;
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (!node) {
+      continue;
+    }
+    // Check in all the callback groups
+    for (auto & weak_group : node->get_callback_groups()) {
+      auto group = weak_group.lock();
+      if (!group || !group->can_be_taken_from().load()) {
+        continue;
+      }
+      group->find_client_ptrs_if([&exec_list](const rclcpp::ClientBase::SharedPtr & client) {
+          if(client){
+            // If any client is found, push it in the exec_list struct
+            exec_list.client.push_back(client);
+            exec_list.number_of_clients++;
+          }
+          return false;
+        });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_waitable_list(ExecutableList & exec_list)
+{
+  // Clear the previous waitables (if any) from the ExecutableList struct
+  exec_list.waitable.clear();
+  exec_list.number_of_waitables = 0;
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (!node) {
+      continue;
+    }
+    // Check in all the callback groups
+    for (auto & weak_group : node->get_callback_groups()) {
+      auto group = weak_group.lock();
+      if (!group || !group->can_be_taken_from().load()) {
+        continue;
+      }
+      group->find_waitable_ptrs_if([&exec_list](const rclcpp::Waitable::SharedPtr & waitable) {
+          if(waitable){
+            // If any waitable is found, push it in the exec_list struct
+            exec_list.waitable.push_back(waitable);
+            exec_list.number_of_waitables++;
+          }
+            return false;
+        });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_executable_list(
+  ExecutableList & executable_list, std::chrono::nanoseconds timeout)
+{
+  // prepare the wait_set
+  prepare_wait_set();
+
+  // add handles to the wait_set and wait for work
+  refresh_wait_set(timeout);
+
+  // Get all the timers
+  get_timer_list(executable_list);
+
+  // Get all the subscribers
+  get_subscription_list(executable_list);
+
+  // Get all the services
+  get_service_list(executable_list);
+
+  // Get all the clients
+  get_client_list(executable_list);
+
+  // Get all the waitables
+  get_waitable_list(executable_list);
+}
+
+void
+StaticSingleThreadedExecutor::execute_ready_executables(
+  ExecutableList & exec_list, std::chrono::nanoseconds timeout)
```

Comment:
Why do we need both the `exec_list` and the `memory_strategy_`?
I think there is some duplication there, as they are mostly collecting the same stuff.

Why not using directly `get_next_ready_executable` instead of re-implementing this?

My idea is that this should look quite similar to the current `SingleThreadedExecutor`, but avoiding calling `clear_handles` and `collect_entities` each time. I don't understand the need of another collection.